synapse/storage/_base.py

```diff
@@ -569,12 +668,56 @@ def _simple_upsert_txn(self, txn, table, keyvalues, values, insertion_values={},
         sql = "INSERT INTO %s (%s) VALUES (%s)" % (
             table,
             ", ".join(k for k in allvalues),
-            ", ".join("?" for _ in allvalues)
+            ", ".join("?" for _ in allvalues),
         )
         txn.execute(sql, list(allvalues.values()))
         # successfully inserted
         return True
 
+    def _simple_upsert_txn_native_upsert(
+        self, txn, table, keyvalues, values, insertion_values={}, best_effort=False
+    ):
+        """
+        Use the native UPSERT functionality in recent PostgreSQL versions.
+        """
+        allvalues = {}
+        allvalues.update(keyvalues)
+        allvalues.update(values)
+        allvalues.update(insertion_values)
+
+        sql = (
+            "INSERT INTO %s (%s) VALUES (%s) "
+            "ON CONFLICT (%s) DO UPDATE SET %s "
+            "RETURNING (xmax = 0) AS inserted"
+        ) % (
+            table,
+            ", ".join(k for k in allvalues),
+            ", ".join("?" for _ in allvalues),
+            ", ".join(k for k in keyvalues),
+            ", ".join(k + "=EXCLUDED." + k for k in values),
+        )
+        try:
+            txn.execute(sql, list(allvalues.values()))
+        except self.database_engine.module.OperationalError as e:
+            # We only care about serialization errors, so check for it
+            if e.args[0] == "could not serialize access due to concurrent update":
+                # A concurrent update problem is when we try and do a native
+                # UPSERT but the row has changed from under us. We can either
+                # retry, or give up if asked to do so.
+                if best_effort:
+                    # If it's a concurrent-update problem, and this is marked as
+                    # 'best effort' (i.e. if there's a race, then the one we
+                    # raced with will suffice), then pretend that we succeeded.
+                    return False
```

Comment:
It's not safe to ignore database errors as the transaction will no longer be valid. This should probably be raised up into `runInteraction` or the like instead.

Reply:
Yup, transactions will be rolled back on error, but they're then effectively closed. Note that the link also says that you need to abort and retry the transaction